### DISCUSSION TITLE
0.5.0-rc1 has been released!

### DISCUSSION THREAD

**Discussion by mihe** [godot-jolt/godot-jolt#526] [Announcements]
This is the first **release candidate** for version 0.5.0, and assuming no major bugs/regressions are found is likely to become 0.5.0-stable.

This release supports [**Godot 4.1**](https://github.com/godotengine/godot/releases/tag/4.1.1-stable).

## Changelog

These are the notable changes that have been made since [0.4.1-stable](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.4.1-stable) was released. You can also find a list of all the commits [here](https://github.com/godot-jolt/godot-jolt/compare/v0.4.1-stable...v0.5.0-rc1).

Any breaking changes are denoted with ⚠️.

### Removed

- ⚠️ Removed the ability to lock all six axes of a `RigidBody3D`. Consider freezing the body as static instead.

### Added

- Added substitutes for all the joint nodes, to better align with the interface that Jolt offers, which consist of `JoltPinJoint3D`, `JoltHingeJoint3D`, `JoltSliderJoint3D`, `JoltConeTwistJoint3D` and `JoltGeneric6DOFJoint3D`. These differ in the following ways:
  - You can enable/disable the limits on all joints.
  - You can enable/disable the joint itself using its `enabled` property.
  - You can fetch the magnitude of the force/torque that was last applied to keep the joint together, using the `get_applied_force` and `get_applied_torque` methods. These coupled with the `enabled` property allows for creating breakable joints.
  - You can increase the joint's solver iterations, to improve stability, using its `solver_velocity_iterations` and `solver_position_iterations` properties.
  - Springs use frequency and damping instead of stiffness and damping.
  - Soft limits are achieved with limit springs.
  - `JoltConeTwistJoint3D` can be configured with a motor.
  - Angular motor velocities are set in radians per second, but displayed in degrees per second.
  - Any motion parameters like bias, damping and relaxation are omitted.
  - Any angular motion parameters for the slider joint are omitted.

### Fixed

- Fixed issue where linear axis locks could be budged a bit if enough force was applied.
- Fixed issue where `CharacterBody3D` and other kinematic bodies wouldn't respect locked axes.
- Fixed issue where passing `null` to the `result` parameter (or omitting it entirely) of the `body_test_motion` method in `PhysicsServer3D` would cause a crash.
- Fixed issue where the `body_is_omitting_force_integration` method in `PhysicsServer3D` would always return `false`.

<hr /><em>This discussion was created from the release <a href='https://github.com/godot-jolt/godot-jolt/releases/tag/v0.5.0-rc1'>0.5.0-rc1</a>.</em>

**Comment by Gamepro5** [comment]
Excellent work!

I have been working on cleaning up the code for my character controller to make it more readable. I'm encountering some odd bugs that I suspect might be collision detection issues as they are not consistent with godot physics. I was wondering if you could tell me what the safe_margin parameter does specifically for the move_and_collide() function, and what the various description-less project settings listed under Godot Jolt do and mean.

**Reply by mihe** [reply]
> I was wondering if you could tell me what the safe_margin parameter does specifically for the move_and_collide() function

I will admit that the `safe_margin` parameter/property is a bit of a mystery to me still, despite having implemented it. Godot's [documentation](https://docs.godotengine.org/en/stable/classes/class_characterbody3d.html#class-characterbody3d-property-safe-margin) for it describes it as being relevant for things like reliably detecting walls and floors, which might very well be true. The reason I had to implement it in the first place (i.e. not just hardcore it to 0 behind the scenes) was because you could otherwise get stuck when walking up certain slopes/ramps.

`move_and_collide` (which is just a thin wrapper for `PhysicsServer3D.body_test_motion`) is internally broken up into three phases. The first phase is the "recovery" phase (depenetration), which moves the body away from any relevant overlaps. The second phase is a shape-cast using the move vector, starting from the "recovered" position. The third and last phase is to then do a collision check at the safe fraction returned from the second phase and return this collision data to the caller, as well as whatever fractions were produced by the second phase.

What's important here is that `safe_margin` is only used during the first (recovery) and last (collision) phase of this sequence, but not the second (cast) phase. It's used as a way to inflate/grow/scale the body by that absolute amount, or giving it what other physics engines might refer to as a "skin". So the end result of that is that you will effectively recover/depenetrate more than what's actually needed, then move the body right up against whatever obstacle you have in front of you, without actually colliding, and finally do a collision check with this extra "skin", effectively guaranteeing a collision.

So to sum up, I believe the purpose of `safe_margin` is to in part help you reliably produce collision data, without needing to actually be in contact with the surface/object, and also to help make the shape-cast actually produce a non-zero safe fraction, since that's typically what's used to actually drive the kinematic body forward.

Also note that `safe_margin` is clamped to a minimum of `0.0001` in both Godot Physics and Godot Jolt, regardless of what you pass into `move_and_collide`.

> what the various description-less project settings listed under Godot Jolt do and mean

As linked in the README (because GDExtension [doesn't support documentation](https://github.com/godot-jolt/godot-jolt/issues/390) yet) you will find the various settings documented [here](https://github.com/godot-jolt/godot-jolt/blob/master/docs/settings.md).

**Reply by MCVenor** [reply]
> move_and_collide (which is just a thin wrapper for PhysicsServer3D.body_test_motion) is internally broken up into three phases. The first phase is the "recovery" phase (depenetration), which moves the body away from any relevant overlaps. The second phase is a shape-cast using the move vector, starting from the "recovered" position. The third and last phase is to then do a collision check at the safe fraction returned from the second phase and return this collision data to the caller, as well as whatever fractions were produced by the second phase.

I was looking for this kind of information or description for a whole day. Thank you!
